src/adapter/src/catalog/storage.rs

```diff
@@ -717,42 +185,55 @@ impl Connection {
         now: NowFn,
         bootstrap_args: &BootstrapArgs,
     ) -> Result<Connection, Error> {
-        // The `user_version` field stores the index of the last migration that
-        // was run. If the upper is min, the config collection is empty.
-        let skip = if is_collection_uninitialized(&mut stash, &COLLECTION_CONFIG).await? {
-            0
+        // Initialize the Stash if it hasn't been already
+        let mut conn = if !stash.is_initialized().await? {
+            // Get the current timestamp so we can record when we booted.
+            let previous_now = mz_repr::Timestamp::MIN;
+            let boot_ts = timeline::monotonic_now(now, previous_now);
+
+            // Initialize the Stash
+            let args = bootstrap_args.clone();
+            stash
+                .with_transaction(move |mut tx| {
+                    Box::pin(async move { stash::initialize(&mut tx, &args, boot_ts.into()).await })
+                })
+                .await?;
+
+            Connection { stash, boot_ts }
         } else {
-            // An advanced collection must have had its user version set, so the unwrap
-            // must succeed.
-            COLLECTION_CONFIG
-                .peek_key_one(&mut stash, USER_VERSION.to_string())
+            // Before we do anything with the Stash, we need to run any pending upgrades.
+            if !stash.is_readonly() {
```

Comment:
Does this readonly check need to be here? I think we should complain loudly if this is the case instead of silently moving on, which is almost certainly wrong (if upgrade needed to happen, then all the dependent types assume that, so not running it will produce a data mismatch between on disk and the binary).

Reply:
I kept the read_only check because it exists today, but you're right, if we need to upgrade and we don't then chances are we'll hit some panic for a data mismatch. 

I got rid of this check and in the upgrade path we'll check if it's read only and we need to upgrade, if so we'll log an error to Sentry.

Reply:
Erp I think I remember why this is needed. Testdrive does a catalog validation on each DDL, and *that* needs to not even try to upgrade things because they'll fail. If it's already there and starts causing test failures, put it back.